### DISCUSSION TITLE
Should showmy.chat support YouTube Live? (and other streaming platforms?)

### DISCUSSION THREAD

**Discussion by BenDMyers** [BenDMyers/showmy.chat#111] [Ideas]
**This is something that I think would need to be a *ways* out there, and would constitute a v2 of the showmy.chat project.**

During today's Frontend Horse stream, someone asked whether showmy.chat supports YouTube Live. Currently, it doesn't… but I don't think it would be entirely out of the question. [`youtube-chat`](https://www.npmjs.com/package/youtube-chat) seems like a good library to use, that would be fairly compatible with our approach to doing things in showmy.chat.

At first blush, it seems like we could add a new route, perhaps something like `showmy.chat/y/UCoc1J1XEEtg4sNrt3V-tNRw`, where that `/y/` indicates a YouTube channel, and the lovely string at the end is a channel ID. This route could be added without impacting the current Twitch chat overlay experience.

However, this opens a can of worms. For instance:

- **Should YouTube overlays leverage the same theme stylesheets as the Twitch overlays, or should we maintain a separate set of stylesheets for YouTube?** I would almost definitely prefer to have one set of theme stylesheets to rule them all — that strikes me as the best fit for ongoing maintenance — but currently, our style hooks (and therefore our stylesheets) are *very* tied to Twitch's API.
- **Should we expose a route for multicasters, such as Restream.io users, who broadcast to multiple platforms at once and would like to show all of their chats at the same time?** This seems a natural conclusion of introducing multiple chat providers, but would not be trivial.

To support anything like this with any degree of flexibility, I think we'd need to rearchitect our chat such that the _core_ of it is platform-agnostic (a unified interface for displaying messages, namely), and treat Comfy as just one of potentially many providers which feeds into that unified platform. Special care would need to be taken to smooth over discrepancies between providers' APIs.

Alternatively… we just say that showmy.chat is Twitch-exclusive and call it a day.

**Comment by ajcwebdev** [comment]
Depends on the goals of the project really, do you want to be the ultimate chat overlay that can be used by any and all streamers for any and all use cases or do you want to be the best overlay for Twitch streamers specifically?

This would definitely open up the field of potential users. For example, this was one of the blockers for using showmy.chat on the StepZen stream since we were broadcasting to both Twitch and YouTube.

The flip side is then you can't be hyper focused on creating the best possible experience for Twitch streamers. Any time spent developing features and integrations with other platforms will necessarily detract from time spent improving the Twitch functionality.

**Reply by BenDMyers** [reply]
This is absolutely a fantastic question and your insight here is wonderful!

I think one thing that might be helpful here would be to find more folks who stream to non-Twitch platforms, as well as people who stream to multiple platforms at once to figure out whether there's even demand here. (Maybe @austingil might be able to weigh in?)

If YouTube-exclusive streamers or multicasters don't feel a need to show their chat as an overlay, I feel like this whole idea would be moot. Worth exploring/investigating, I think!

**Reply by AustinGil** [reply]
I multi-cast, but I'm at a point right now where I feel like focusing in more on JUST Twitch. So although it would be awesome to have chat supported for at least these two platforms, it's not a deal breaker for me. I need to start adding some chat overlays to my stream, and I dont think I need YT chat.

¯\(°_o)/¯

**Comment by NickyMeuleman** [comment]
So I spent a bit of time on looking what's out there relating to this and here is what I found:

Showing chat on youtube livestreams in significantly less popular than on Twitch.
However, the demand for (a simple way) to show chat onscreen seems to be very large (if you ask if people want it, the answer is a resounding "yes").

It can be done right now, however, it's clunky and unintuitive.
The most userfriendly method I found is to first pop out the chat for a livestream, copy the URL, use that as browsersource.
This URL changes every time you go live.
Another method is shown [in a TVN video](https://www.youtube.com/watch?v=bwbxbuxpPN4). It's ... not pretty.
Conclusion: I absolutely believe that if showmy.chat decides to support multiple platforms, it can become the goto method.

By default for those youtube methods, you get the standard look, which is probably not what streamers want if they're going to the trouble of showing the chat on-stream.
You can add some custom CSS in OBS to customize it.
The most userfriendly stylesheet generator I found is https://chatv2.septapus.com/
A very similar concept to showmy.chat right now.

The can of worms point: I 100% agree. Every possible method of supporting multiple sources I could think of comes with significant downsides.
What is similar in 2 sources (yt and twitch) today, might diverge (slightly?) tomorrow.

The possibility I thought about most is to have a deliberately small platform agnostic core that different sources can then "enrich".
This probably takes the form of a Twitch version pulling data via comfy.js and making DOM nodes that can then be styled by CSS. Similar for different platforms. pull in data, then create DOM elements that can be styled by a theme.

As @BenDMyers also mentioned, this also means that those themes unfortunately are very tied to a specific platform.
As attractive as sharing the whole stylesheet between multiple platforms is, I don't think it'll be feasable longterm, as that means supporting everything in one place, with all possible current and future conflics (naming, specificity, overrides, having things that have to be structured one way for service A, but are subtly different in service B, ...). Also, there is little point in shipping a twitch specific feature to a youtube-only stream.

I think you hit the nail on the head there at the end.
A platform-agnostic core that different parts can plug into.
A bit like how Vue is architected (@fimion's explanation of this on a recent stream was wonderful).
Those parts can be:
- A way to pull in data from that platform.
- A way to turn that data into DOM elements.
- A way to style those DOM elements.

That way, the supporting multistreaming becomes way more straightforward too, plug in multiple parts, and add a part that makes them all play nice.
On the subject of those themes: maybe themes have a core file that has most styling rules and adds a platform specific css file? 

edit: I'd also like to mention how important I personally think whatever we go with remains: "easy to start, possible to customize".

This includes the option where showmy.chat remains Twitch exclusive.

**Comment by JoshuaKGoldberg** [comment]
FWIW I would _really_ like this. Having a unified chat zone is roughly the only thing blocking me from live streaming to both Twitch and YouTube (since I stream with raw OBS and can open two instances). Streaming straight to YouTube would be really nice - today I manually download videos from Twitch & upload them to YouTube...

@BenDMyers how can I/we help make this happen? 😁

**Comment by AustinGil** [comment]
Yeah, coming back to this, Twitch has kind of soured on me, so anything that would make it easier to move away would be awesome :)
